### ISSUE TITLE
next/link sometimes does not update the page when used in dynamic page with replace and query parameters in v15

### ISSUE THREAD

**Issue by skirsten** [vercel/next.js#69847]
### Link to the code that reproduces this issue

https://github.com/skirsten/repro-next-link-broken

### To Reproduce

1. Start the repo with `next build` and `next start`! It is not reproducible with `next dev`.
2. Read the text and click the links in the order that is explained on the page.
3. After each click check the url and the first line of the output. You should see the problem.

Alternatively use this hosted url: https://repro-next-link-broken.vercel.app/
Or checkout this video: https://github.com/user-attachments/assets/bfcee9ae-e95e-405e-89a1-83e8d4256538 (if you turn up the volume you can hear the mouse clicks)

### Current vs. Expected behavior

I expect the links to work as they did in next v14 without problems.

Known working:
14.2.5

Not working in at least:
15.0.0-canary.144
15.0.0-canary.136
15.0.0-canary.128

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #40~22.04.3-Ubuntu SMP PREEMPT_DYNAMIC Tue Jul 30 17:30:19 UTC 2
  Available memory (MB): 32012
  Available CPU cores: 16
Binaries:
  Node: 20.12.1
  npm: 10.5.0
  Yarn: 1.22.22
  pnpm: 9.9.0
Relevant Packages:
  next: 15.0.0-canary.145 // Latest available version is detected (15.0.0-canary.145).
  eslint-config-next: 15.0.0-canary.145
  react: 19.0.0-rc-7771d3a7-20240827
  react-dom: 19.0.0-rc-7771d3a7-20240827
  typescript: 5.5.4
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Navigation

### Which stage(s) are affected? (Select all that apply)

next start (local), Vercel (Deployed), Other (Deployed)

### Additional context

Unfortunately the nature of this bug is not very exact and sometimes changing the layout or order of the links changes how they will behave.

**Comment by ztanner** [comment]
Thank you for the very clear reproduction, I'll have a fix up for this shortly.

**Comment by skirsten** [comment]
Thank you for the quick response and fix. I checked out the latest canary and it seems most issues are fixed :tada:.

Unfortunately I encountered in some cases (depending on the order of the links) it still requires clicking the link twice.

I added a new reproduction here https://github.com/skirsten/repro-next-link-broken/tree/bug2 and hosted here: https://repro-next-link-broken-git-bug2-skirsten.vercel.app/.

I believe its related to the same logic.

**Comment by ztanner** [comment]
> Thank you for the quick response and fix. I checked out the latest canary and it seems most issues are fixed 🎉.
> 
> Unfortunately I encountered in some cases (depending on the order of the links) it still requires clicking the link twice.
> 
> I added a new reproduction here https://github.com/skirsten/repro-next-link-broken/tree/bug2 and hosted here: https://repro-next-link-broken-git-bug2-skirsten.vercel.app/.
> 
> I believe its related to the same logic.

Indeed! Thanks for flagging, I’ll be taking a look at this soon.

**Comment by ztanner** [comment]
Hi @skirsten -- I believe this should be resolved by #69948 and added your reproduction as a regression test. I've just triggered a canary. Appreciate the reports, let me know if you run into any troubles!

**Comment by skirsten** [comment]
Awesome, I tested the canary and all the links behave as expected now. Thanks :tada:

I did notice one small inconsistency where some client side cache is used unexpectedly.
Its hard to explain so I put a small reproduction together:
https://github.com/skirsten/repro-next-link-broken/tree/bug3
https://repro-next-link-broken-git-bug3-skirsten.vercel.app/

Basically when returning to page 1 for the first time, its reusing the old page 1 (the timer will not start at 0).
